### PR TITLE
Blob storage helper methods

### PR DESCRIPTION
**Pull request by yiwu-arbug**
Split out interfaces needed for blob storage from #1560, including
* CompactionEventListener and OnFlushBegin listener interfaces.
* Blob filename support.

Test Plan:
make all check

### CODE REVIEW

File: db/compaction_job.cc [hunk 1]
```diff
@@ -739,12 +739,21 @@ void CompactionJob::ProcessKeyValueCompaction(SubcompactionState* sub_compact) {
     input->SeekToFirst();
   }
 
+  // we allow only 1 compaction event listener. Used by blob storage
+  std::shared_ptr<CompactionEventListener> comp_event_listener = nullptr;
```
Comment (siying) [inline db/compaction_job.cc]: Why not unique_ptr?

File: db/compaction_iterator.cc [hunk 2]
```diff
@@ -235,6 +266,11 @@ void CompactionIterator::NextFromInput() {
         }
       }
     } else {
+      if (compaction_listener_)
+        compaction_listener_->OnCompaction(compaction_->level(), ikey_.user_key,
+                                           fromInternalValueType(ikey_.type),
+                                           value_, ikey_.sequence, false);
```
Comment (siying) [inline db/compaction_iterator.cc]: Maybe add { } too.

File: db/compaction_job.cc (hunk 1 above)
Comment (yiwu-arbug) [inline db/compaction_job.cc]: I'm following DBOptions.listeners, which stores shared_ptr.

Comment (siying) [inline db/compaction_job.cc]: Can we follow the pattern like CompactionFilterFactory and TableFactory?

Comment (yiwu-arbug) [inline db/compaction_job.cc]: Then probably just raw pointers.

**Pull request closed without merging**